src/Core/Form/IdentifiableObject/FormHandler.php

```diff
@@ -0,0 +1,242 @@
+     */
+    private $dataHandler;
+
+    /**
+     * @var FormFactoryInterface
+     */
+    private $formFactory;
+
+    /**
+     * @var HookDispatcherInterface
+     */
+    private $hookDispatcher;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var bool
+     */
+    private $isDemoModeEnabled;
+
+    /**
+     * @param string $formType
+     * @param FormDataProviderInterface $dataProvider
+     * @param FormDataHandlerInterface $dataHandler
+     * @param FormFactoryInterface $formFactory
+     * @param HookDispatcherInterface $hookDispatcher
+     * @param TranslatorInterface $translator
+     * @param bool $isDemoModeEnabled
+     */
+    public function __construct(
+        $formType,
+        FormDataProviderInterface $dataProvider,
+        FormDataHandlerInterface $dataHandler,
+        FormFactoryInterface $formFactory,
+        HookDispatcherInterface $hookDispatcher,
+        TranslatorInterface $translator,
+        $isDemoModeEnabled
+    ) {
+        $this->formFactory = $formFactory;
+        $this->dataProvider = $dataProvider;
+        $this->dataHandler = $dataHandler;
+        $this->hookDispatcher = $hookDispatcher;
+        $this->translator = $translator;
+        $this->formType = $formType;
+        $this->isDemoModeEnabled = $isDemoModeEnabled;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getForm(array $options = [])
+    {
+        return $this->buildForm(
+            $this->dataProvider->getDefaultData(),
+            null,
+            $options
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFormFor($id, array $options = [])
+    {
+        return $this->buildForm(
+            $this->dataProvider->getData($id),
+            $id,
+            $options
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(FormInterface $form)
+    {
+        return $this->handleForm($form);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handleFor($id, FormInterface $form)
+    {
+        return $this->handleForm($form, $id);
+    }
+
+    /**
+     * @param mixed $data
+     * @param int|null $id
+     * @param array $options
+     *
+     * @return FormInterface
+     */
+    private function buildForm($data, $id = null, array $options = [])
+    {
+        $formBuilder = $this->formFactory->createBuilder($this->formType, $data, $options);
+
+        $this->hookDispatcher->dispatchWithParameters('action' . $formBuilder->getName() . 'FormBuilderModifier', [
+            'form_builder' => $formBuilder,
+            'data' => &$data,
+            'id' => $id,
+        ]);
+
+        return $formBuilder->getForm();
+    }
+
+    /**
+     * @param FormInterface $form
+     * @param int|null $id
+     *
+     * @return int
+     */
+    private function handleForm(FormInterface $form, $id = null)
+    {
+        if (!$form->isSubmitted()) {
+            return 0;
```

Comment:
im still trying to figure out a better return value, any suggestions?

Reply:
I guess you mean "what should I return if there are errors in the form ?" ?

Ideas:
- return 0, but as a constant so we can use some test like `$result = $handler->handle($form); if ($result === FormHandlerInterface::ERRORS_FOUND) {...}`
- return a dedicated exception `CannotHandleFormException` which contains the array of errors, and catch it in the controller
- return a `FormHandleResult` value object with a `$hasErrors` property or even a `hasErrors()` functiton

Reply:
actually if form is not submitted, its fine, we just skip its handling (since `handle()` is called for both GET and POST methods). throwing exception is also not a best idea i think, as no unexpected situation or anything wrong happened here. so im thinking about some VO, wdyt?

Reply:
even though I'm not a big fan, maybe a `FormHandleResult` would do the job, this the only way I see to be able to return all the informations: success|invalid|skipped|errors
the only other way would for the `FormHandler` to keep a reference on the form and have methods to return those values `isValid`, `isSubmitted`, ... But it is already accessible vie the $form object 🧐
(sorry I'm thinking out loud 😄)

Reply:
actually if the only job of the `FormHandler` was to handle the form submit/validation it would be easier, but we would need to change its prototypes a bit:

```php
     /**
       * @var array $data is an array, this way we don't bind it to the Request object (it' the controller job to extract the request params, or another object in case of too complicated transformations)
       * @return bool
      **/
     public function handleForm(FormInterface $form, array $data)
     {
             //Hook to modify the data before setting it to the form
             $form->submit($data);
             //Check demo mode
             //Check isSubmit
             //Check isValid
             //return false when not valid
            
             //If all is good then we call the form creation/update
             //return true if all is good
             //the errors are stored in the Form as it is done here
     }
```

Reply:
@jolelievre we cant just return `true` or `false`, because `FormHandler` delegates create/update to other services and return `id` result on success.

Reply:
ok, sorry
then maybe return false on failure, rather than 0, and the ID on success?

Reply:
yep `null` is fine for me to
but what do mean by `?int`

Reply:
OK for integer or Value Object 😉 

> actually if form is not submitted

We had a bug once in the Product page because form was not submitted 🤔 ... so let's keep this as it is now, but be wary of it when we start using it